build_test/lib/src/in_memory_reader.dart

```diff
@@ -9,29 +9,39 @@ import 'package:glob/glob.dart';
 
 import 'in_memory_writer.dart';
 
+/// An [AssetReader] that records which assets have been read to [readAssets].
+abstract class RecordingAssetReader implements AssetReader {
+  Iterable<AssetId> get readAssets;
+}
+
 /// An implementation of [AssetReader] with primed in-memory assets.
-class InMemoryAssetReader implements MultiPackageAssetReader {
+class InMemoryAssetReader
+    implements MultiPackageAssetReader, RecordingAssetReader {
   final Map<AssetId, DatedValue> assets;
   final String rootPackage;
+  final Set<AssetId> readAssets;
 
   /// Create a new asset reader that contains [sourceAssets].
   ///
   /// May optionally define a [rootPackage], which is required for some APIs.
   InMemoryAssetReader({Map<AssetId, DatedValue> sourceAssets, this.rootPackage})
-      : assets = sourceAssets ?? <AssetId, DatedValue>{};
+      : assets = sourceAssets ?? <AssetId, DatedValue>{},
+        readAssets = new Set<AssetId>();
 
   @override
   Future<bool> canRead(AssetId id) async => assets.containsKey(id);
```

Comment:
nit: we should probably add these assets as well, calling `canRead` on an asset adds a dependency on it just like `readAs*`

Reply:
Done.

Reply:
Just to clarify, does this mean that you would expect there to be a dependency on a file that returns false from canRead? In an offline conversation, I gathered that this is true.